File: releases/LoopFlow_Rhino-to-Octane-Sync/Python/LiveLink_R2O_Camera.py
# -*- coding: utf-8 -*-
"""
============================================================
Script Name        : LiveLink Rhino to Octane Standalone (Camera Sender)
Version            : v1.0
Date               : 2026-04-28
Author             : Cursor + Claude Sonnet 4.6
Environment        : Rhino 8 (CPython 3.9) / Python 3
Sync File          : Determined by the CameraFile field in R2O_Path.txt (default: R2O_Camera_Sync_Data.lua)
============================================================
[Usage]
1. Environment: the script auto-reads or creates the config file
   (%APPDATA%\McNeel\Rhinoceros\8.0\scripts\LoopFlow_R2O\Data\R2O_Path.txt).
2. Live sync: no need to save the file first; new unsaved files are synced immediately.
3. Run the script to start/stop camera live sync.

[Variable Notes]
- Reads R2O_Path.txt:
  - `DataPath`  : output directory for the sync file
  - `CameraFile`: sync file name (default: R2O_Camera_Sync_Data.lua)
- Exceptions are written to: %APPDATA%\McNeel\Rhinoceros\8.0\scripts\LoopFlow_R2O\Data\cursor_R2O_debug_log.txt
============================================================
"""
import Rhino
import scriptcontext as sc
import os
import math
import sys
import System
import time

_HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, _HERE)
from LiveLink_R2O__Config import load_r2o_config, log_exception, safe_write_text_atomic

# ── Module-level constants (centralised for easy tuning) ──────────────────
DEFAULT_LENS_MM  = 50.0
MIN_INTERVAL_SEC = 0.2
EPS_POSITION     = 1e-4
EPS_FOV          = 1e-4

def export_octane_camera(sender, e):
    try:
        doc = Rhino.RhinoDoc.ActiveDoc
        if not doc:
            return

        if not doc.Views.ActiveView:
            return

        vp = doc.Views.ActiveView.ActiveViewport

        cfg = load_r2o_config()
        export_dir = cfg["DataPath"]
        sync_file_name = cfg["CameraFile"]
        sync_file_path = os.path.join(export_dir, sync_file_name)

        scale = Rhino.RhinoMath.UnitScale(doc.ModelUnitSystem, Rhino.UnitSystem.Meters)

        def convert_coord(rh_obj, is_point=True):
            oct_x = rh_obj.X
            oct_y = rh_obj.Z
            oct_z = -rh_obj.Y
            if is_point:
                return oct_x * scale, oct_y * scale, oct_z * scale
            return oct_x, oct_y, oct_z

        loc = convert_coord(vp.CameraLocation, True)
        tar = convert_coord(vp.CameraTarget, True)
        up = convert_coord(vp.CameraUp, False)

        lens_length = float(vp.Camera35mmLensLength) if vp.Camera35mmLensLength else 0.0
        if lens_length <= 0.0:
            lens_length = DEFAULT_LENS_MM

        fov_degrees = 2.0 * math.atan(36.0 / (2.0 * lens_length)) * (180.0 / math.pi)

        payload = (loc, tar, up, round(fov_degrees, 6))

        state_key = "R2O_CAMERA_SYNC_STATE_V3_2"
        state = sc.sticky.get(state_key, {})
        now = time.monotonic()

        last_time = state.get("last_time", 0.0)
        last_payload = state.get("last_payload")

        if last_payload is not None:
            try:
                loc_d = max(abs(payload[0][i] - last_payload[0][i]) for i in range(3))
                tar_d = max(abs(payload[1][i] - last_payload[1][i]) for i in range(3))
                up_d = max(abs(payload[2][i] - last_payload[2][i]) for i in range(3))
                fov_d = abs(payload[3] - last_payload[3])
                no_change = loc_d < EPS_POSITION and tar_d < EPS_POSITION and up_d < EPS_POSITION and fov_d < EPS_FOV
            except Exception:
                no_change = False

            if no_change:
                return

        if (now - last_time) < MIN_INTERVAL_SEC:
            return

        lua_content = """return {{
    position = {{{0:.5f}, {1:.5f}, {2:.5f}}},
    target = {{{3:.5f}, {4:.5f}, {5:.5f}}},
    up_vector = {{{6:.5f}, {7:.5f}, {8:.5f}}},
    fov_degrees = {9:.5f}
}}""".format(
            loc[0], loc[1], loc[2],
            tar[0], tar[1], tar[2],
            up[0], up[1], up[2],
            fov_degrees
        )

        safe_write_text_atomic(sync_file_path, lua_content, encoding="utf-8")

        state["last_time"] = now
        state["last_payload"] = payload
        state["sync_file_path"] = sync_file_path
        sc.sticky[state_key] = state
    except Exception as exc:
        log_exception("LiveLink_R2O_Camera", exc)

def toggle_octane_camera_sync():
    event_key = "R2Octane_Camera_Sync_Delegate_V3_2"

    if event_key in sc.sticky:
        handler = sc.sticky[event_key]
        Rhino.Display.RhinoView.Modified -= handler
        sc.sticky.pop(event_key, None)
        print("R2O Camera Sync: Live sync stopped.")
    else:
        cfg = load_r2o_config()
        handler = System.EventHandler[Rhino.Display.ViewEventArgs](export_octane_camera)
        sc.sticky[event_key] = handler
        Rhino.Display.RhinoView.Modified += handler

        export_octane_camera(None, None)
        print("R2O Camera Sync: Live sync started. (Output bound to {})".format(cfg["DataPath"]))

if __name__ == "__main__":
    toggle_octane_camera_sync()

File: releases/LoopFlow_Rhino-to-Octane-Sync/LUA/LiveLink_R2O_Camera.lua
-- ============================================================
-- Script Name  : LiveLink_R2O_Camera
-- Version           : v1.0
-- Date              : 2026-04-28
-- Author            : Cursor + Claude Sonnet 4.6
-- Description : Auto-parses the config file
--               (%APPDATA%\McNeel\Rhinoceros\8.0\scripts\LoopFlow_R2O\Data\R2O_Path.txt),
--               reads the camera sync file, and syncs the Rhino viewport to Octane in real time.
-- ============================================================

-- @description Rhino to Octane Camera Sync Macro
-- @shortcut Ctrl + Q
--
-- [Usage]
-- 1) Enable `LiveLink_R2O_Camera.py` on the Rhino side (keep it running).
-- 2) In Octane Standalone, press `Ctrl + Q` or run this script to apply the latest camera sync.
-- 3) Note: the Thin Lens Camera node in the Octane scene must be "Expanded" out of the
--    Render Target as a standalone node; a collapsed Camera cannot be accessed by this script.
--
-- [Variable Notes]
-- - Reads the DataPath and CameraFile fields from R2O_Path.txt.

-- ── Path constants (derived automatically from install location, no hard-coding) ──────────
local APPDATA     = os.getenv("APPDATA")
local INSTALL_DIR = APPDATA .. "\\McNeel\\Rhinoceros\\8.0\\scripts\\LoopFlow_R2O"
local DATA_DIR    = INSTALL_DIR .. "\\Data"
local CONFIG_FILE = DATA_DIR .. "\\R2O_Path.txt"

local function loadConfig()
    local cfg = {
        DataPath   = DATA_DIR,
        CameraFile = "R2O_Camera_Sync_Data.lua",
    }
    local f = io.open(CONFIG_FILE, "r")
    if f then
        for line in f:lines() do
            local k, v = line:match("^(%w+):%s*(.*)")
            if k and cfg[k] ~= nil then cfg[k] = v end
        end
        f:close()
    end
    return cfg
end

local function getThinLensCamera()
    local graph = octane.project.getSceneGraph()
    -- Camera must be "Expanded" out of the Render Target as a standalone node to be found;
    -- a collapsed Camera cannot be accessed via the Octane Lua API.
    local cams = graph:findNodes(octane.NT_CAM_THINLENS, true)
    if cams and #cams > 0 then return cams[1] end
    return nil
end

local function main()
    local cfg = loadConfig()
    local dataPath = cfg.DataPath:match("^%s*(.-)%s*$"):gsub("\\", "/")
    local syncPath = dataPath .. "/" .. cfg.CameraFile:match("^%s*(.-)%s*$")

    local chunk, err = loadfile(syncPath)
    if not chunk then
        print("[Error] Camera sync file not found: " .. syncPath)
        return
    end

    local status, data = pcall(chunk)
    if not status or type(data) ~= "table" then
        print("[Error] Failed to parse camera data.")
        return
    end

    local camNode = getThinLensCamera()
    if not camNode then
        print("[Warning] Thin Lens Camera node not found. Please ensure at least one camera exists in the scene.")
        return
    end

    print("========================================")
    camNode:setPinValue(octane.P_POSITION, data.position)
    camNode:setPinValue(octane.P_TARGET, data.target)
    camNode:setPinValue(octane.P_UP, data.up_vector)
    camNode:setPinValue(octane.P_FOV, data.fov_degrees)
    print("[Success] Camera view synced successfully.")
    print("========================================")
end

main()
